src/Analyser/MutatingScope.php

```diff
@@ -3803,11 +3806,32 @@ public function filterBySpecifiedTypes(SpecifiedTypes $specifiedTypes): self
 			return $b['sure'] - $a['sure']; // @phpstan-ignore-line
 		});
 
-		$scope = $this;
 		$specifiedExpressions = [];
 		foreach ($typeSpecifications as $typeSpecification) {
 			$expr = $typeSpecification['expr'];
 			$type = $typeSpecification['type'];
+
+			if ($expr instanceof NotIssetExpr) {
+				$unsetExpr = $expr->getExpr();
+				$scope = $scope->unsetExpression($unsetExpr);
+
+				if ($unsetExpr instanceof Expr\ArrayDimFetch) {
+					$type = $scope->getType($unsetExpr->var);
+					$arraySize = $type->getArraySize();
+
+					if (
+						$arraySize instanceof ConstantIntegerType
+						&& $arraySize->getValue() === 0
+					) {
+						$exprString = $this->getNodeKey($unsetExpr->var);
+
+						unset($scope->expressionTypes[$exprString]);
+						unset($scope->nativeExpressionTypes[$exprString]);
```

Comment:
I don't understand this logic at all. Here you're saying that if we `unset($var['foo'])` and the `$var` is an empty array, you're going to unset `$var` !?

Reply:
this is the implementation part of [the tests which fail atm](https://github.com/phpstan/phpstan-src/pull/2657#issuecomment-1747523369)

the idea is to implement [what was described](https://github.com/phpstan/phpstan-src/pull/2657#pullrequestreview-1651857463) in

> Because for $foo with certainty maybe and not-nullable, the only !isset($foo) is that the variable does not exist, so the certainty needs to be no.

let me fix it

Reply:
I am doing this array stuff here, because I think:

a array shape which only has a single offset, which gets checked with isset() on said offset will  never have the variable defined in the falsey-branch

```php
$a = ['foo' => 'abc];
if (isset($a['foo'])) {
} else {
  // $a is always undefined 
}

```

Reply:
since we don't want to go from YES to NO certainty,  I think my above case is not possible and I should not try to support it?

Reply:
```php
$a = ['foo' => 'abc'];
if (isset($a['foo'])) {
} else {
  // $a should be NeverType here
}
```